### DISCUSSION TITLE
IAsyncDisposable not working

### DISCUSSION THREAD

**Discussion by truthz03** [dotnet/aspnetcore#25817] [General] [answered]
Hi,

I have a component which inherits from OwningComponentBase.
If I change the page all scoped services from this component disposes.

So if I implement the IDisposable inside one of my scoped services everything is ok.

My problem is that I want to save data before disposing the service.
But If I implement IAsyncDisposable insead of IDisposable inside my service I get an error:

```
blazor.webassembly.js:1 crit: Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: 'XYZ.Client.Services.Objectsecurity.ObjectsecurityService' type only implements IAsyncDisposable. Use DisposeAsync to dispose the container.
System.InvalidOperationException: 'XYZ.Client.Services.Objectsecurity.ObjectsecurityService' type only implements IAsyncDisposable. Use DisposeAsync to dispose the container.
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.Dispose()
   at Microsoft.AspNetCore.Components.OwningComponentBase.System.IDisposable.Dispose()
   at Microsoft.AspNetCore.Components.Rendering.ComponentState.TryDisposeInBatch(RenderBatchBuilder batchBuilder, Exception& exception)
```

Is IAsyncDisposable not supported yet?
What can I do alternatively?

Thanks Thomas

**Comment by SteveSandersonMS** [comment, marked as answer]
We didn't yet add `IAsyncDisposable` support to `OwningComponentBase`, but you can implement it manually on your subclass.

In your `.razor` component that has `@inherits OwningComponentBase<TSomething>`, add:

```razor
@implements IAsyncDisposable
```

Then in your component's `@code` block, also add:

```cs
public ValueTask DisposeAsync() => Service.DisposeAsync();
```

I'm filing https://github.com/dotnet/aspnetcore/issues/25873 to track improving this.
